### PR TITLE
Add icdf functions for Beta, Gamma, Chisquared and StudentT distributions

### PR DESCRIPTION
**Pull request by amyoshino**
**What is this PR about?**
Adds ICDF functions to Beta, Gamma, Chisquared and StudentT distributions.

Issue https://github.com/pymc-devs/pymc/issues/6612

This is a work in progress. The quantile functions depend on two main functions: 
- Inverse of the regularized incomplete beta function: `betaincinv`
- Inverse to the regularized lower incomplete gamma function: `gammaincinv`

Those are implemented in SciPy but not in Pytensor yet. When importing the two functions from `scipy.special` and plugging them in the desired formula, I get a `TypeError` like this below: 

`"TypeError: ufunc 'betaincinv' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''"`

image

It looks like the most straightforward way to make these icdf functions work is to implement them in Pytensor, in the same way the [`pt.betainc` has been implemented](https://github.com/pymc-devs/pytensor/blob/0b558d8e65ee6603b80be566d36a036b68a97ee0/pytensor/scalar/math.py#L1335)

Any comments about if this is the right path? If so I am happy to open an issue in Pytensor as a feature request, and possibly helping on its development if nobody else wants to take it. But if there is any ideas on how to make it simpler than that, let me know. 

References:
- SciPy betaincinv: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.betaincinv.html
- SciPy gammaincinv: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammaincinv.html
- ChiSquared: https://www.wolframalpha.com/input?i=chi+squared+distribution+quantile+function
- Beta: https://www.wolframalpha.com/input?i=beta+distribution+quantile+function
- Gamma: https://www.wolframalpha.com/input?i=gamma+distribution+quantile+function
- StudentT: https://www.wolframalpha.com/input?i=student+distribution+quantile+function

**Checklist**
+ [ ] Explain important implementation details 👆
+ [ ] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [ ] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [ ] Are the changes covered by tests and docstrings?
+ [ ] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## New features
- ...

## Bugfixes
- ...

## Documentation
- ...

## Maintenance
- ...

----
:books: Documentation preview :books:: https://pymc--6845.org.readthedocs.build/en/6845/

### CODE REVIEW

**Comment by amyoshino** [comment]
@ricardoV94 now that https://github.com/pymc-devs/pytensor/pull/502 is merged, we are able to complete this PR.

The tests are failing because `betaincinv` and `gammaincinv` cannot be imported from `pytensor.tensor.math`, I guess it is matter of time to pytensor to be updated. I will restart the checks when it happens. 

Let me know what you think.

**Comment by ricardoV94** [comment]
They should be imported directly from pytensor.tensor.

pymc.math has a subset of hand-picked functions for users

**Comment by ricardoV94** [comment]
You can see which PyTensor version is being installed in the CI.

File: pymc/distributions/continuous.py
Comment (ricardoV94) [inline pymc/distributions/continuous.py]: Should use https://github.com/pymc-devs/pymc/blob/main/pymc%2Fdistributions%2Fdist_math.py#L76

And in the other functions 

This is because the default `check_parameters` was developed for logp in mind and pymc replaces it by -inf sometimes, which doesn't make sense for icdf expressions

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pymc/distributions/continuous.py
Comment (ricardoV94) [inline pymc/distributions/continuous.py]: This is useless?

File: pymc/distributions/continuous.py
Comment (ricardoV94) [inline pymc/distributions/continuous.py]: If beta is just used for the assert, could instead assert scale and avoid the reciprocal

File: pymc/distributions/continuous.py
Comment (ricardoV94) [inline pymc/distributions/continuous.py]: Not needed. This Class just returns a Gamma variable so the icdf will work already like this.

File: tests/distributions/test_continuous.py
Comment (ricardoV94) [inline tests/distributions/test_continuous.py]: No need to test, the Gamma is enough

File: tests/distributions/test_continuous.py
Comment (ricardoV94) [inline tests/distributions/test_continuous.py]: No need for both parametrization tests. That's something we used to do for logp but makes no sense really, unless the expression we used depended on the parametrization

**Review by ricardoV94 — COMMENTED** [review]
Couple more comments. Looks great otherwise

**Comment by ricardoV94** [comment]
Yeah it's strange it's still installing 2.18.4, even though 2.18.5 was released already some days ago. We may need to be just a bit more patient?

File: pymc/distributions/continuous.py
Comment (amyoshino) [reply pymc/distributions/continuous.py]: Ops, my bad! Actually I just realized that some of the other icdf functions I have implemented were using `check_parameters` instead of `check_icdf_parameters`. I've changed them in the commit https://github.com/pymc-devs/pymc/pull/6845/commits/be24e627a698659037db8cb8affb9c82ad24d2ee, but if you think those are out of topic, I can revert it and create a new PR only for these small changes.

File: pymc/distributions/continuous.py
Comment (amyoshino) [reply pymc/distributions/continuous.py]: It looks like so, I kept it earlier because the logcdf function above it also uses it in the same way. 
Because we are testing from tau as well in the test suite, I thought it would be part of the tau/sigma conversion, or that it initialize sigma = 1 whenever neither sigma and tau are not passed as arguments. 
However, when I take it out, the tests pass anyways even for non initialized arguments, so I think it is useless.
Removed in the commit https://github.com/pymc-devs/pymc/pull/6845/commits/be24e627a698659037db8cb8affb9c82ad24d2ee

File: pymc/distributions/continuous.py
Comment (amyoshino) [reply pymc/distributions/continuous.py]: Done! Even though checking for scale, I kept beta on the user facing message, because this is the parameterization used in the documentation. Fixed in https://github.com/pymc-devs/pymc/pull/6845/commits/be24e627a698659037db8cb8affb9c82ad24d2ee

File: pymc/distributions/continuous.py
Comment (amyoshino) [reply pymc/distributions/continuous.py]: Very interesting!! Removed that! Thanks for pointing this out. 
Fixed in https://github.com/pymc-devs/pymc/pull/6845/commits/be24e627a698659037db8cb8affb9c82ad24d2ee

File: tests/distributions/test_continuous.py
Comment (amyoshino) [reply tests/distributions/test_continuous.py]: Forgot to change it in https://github.com/pymc-devs/pymc/commit/be24e627a698659037db8cb8affb9c82ad24d2ee. Fixed in https://github.com/pymc-devs/pymc/pull/6845/commits/fdfe4a1de8dc59293e8ab1e935780f5f7257f2e3.

File: tests/distributions/test_continuous.py
Comment (amyoshino) [reply tests/distributions/test_continuous.py]: Pardon my lack of attention, I forgot to add this as well in previous commits. Fixed in https://github.com/pymc-devs/pymc/pull/6845/commits/e045748f94a51bf1d7905ff7526bbe82f03577e9

**Comment by amyoshino** [comment]
> Yeah it's strange it's still installing 2.18.4, even though 2.18.5 was released already some days ago. We may need to be just a bit more patient?

Right! That's what I thought, I saw the pytenor release 4 days ago and thought it was a good experience to try it now and see if the update immediately applies. No rush on it, it will work soon.

File: pymc/distributions/continuous.py
Comment (ricardoV94) [reply pymc/distributions/continuous.py]: It's fine to do it here. Thanks for the fix

File: pymc/distributions/continuous.py
Comment (ricardoV94) [reply pymc/distributions/continuous.py]: Sounds good

File: pymc/distributions/continuous.py
Comment (ricardoV94) [reply pymc/distributions/continuous.py]: Can you just squash the commits related to the new functionality into a single commit and have the fix for the previous implementations as a separate commit?

You will have to force-push after that. Also if it's your first time backup your branch first!

Comment (amyoshino) [reply pymc/distributions/continuous.py]: I will try that! It will be an interesting experience.

Comment (amyoshino) [reply pymc/distributions/continuous.py]: After failing quite a bit, it is done!
Icdf implementations + tests in https://github.com/pymc-devs/pymc/pull/6845/commits/babc47923fd8272014be2c2f948ad917cfc60609
Fixing `check_parameters` --> `check_icdf_parameters` in https://github.com/pymc-devs/pymc/pull/6845/commits/c54cd1d05cfe9fabd913bcf1479dc6c730cd53b6

Comment (ricardoV94) [reply pymc/distributions/continuous.py]: Awesome. I see however there's a 3rd commit: "merge main". usually when keeping a clean git history in your PR you would update via pull rebase instead of the default pull merge.

I think this won't be a problem and the commit will drop when we rebase and merge but something I thought you might want to know

**Comment by ricardoV94** [comment]
@maresb any idea why CI is not yet installing the last PyTensor? Was there an issue on the conda side?

File: pymc/distributions/continuous.py
Comment (amyoshino) [reply pymc/distributions/continuous.py]: Thanks for guiding me through this. I always used the "update branch" in the github UI at the top of the repository, whenever I start working or before the revision. 
The previous "merge main" was kind of a hassle when I tried to rebase to squash commits. It was a good learning lesson, and from now on I will shift to pull rebase.

Comment (ricardoV94) [reply pymc/distributions/continuous.py]: I'm still learning myself. The recommended source, if you want to learn it less randomly, is usually https://git-scm.com/book/en/v2

Some chapters are more useful than others if you already know you are working specifically with git + github

Comment (amyoshino) [reply pymc/distributions/continuous.py]: Nice!! Thank you for the resource! I definitely want to level up in git skills.

**Comment by maresb** [comment]
It uploaded successfully on anaconda.org, so I restarted the tests.

**Comment by amyoshino** [comment]
Interesting way to restart the tests! I searched for and the recommendation I found was to push an empty commit, but I think this is nicer.

**Comment by ricardoV94** [comment]
You can also just ask a dev woth write permissions. We can restart them from GitHub interface

**Comment by amyoshino** [comment]
Apologies, I have re-requested review by mistake.

**Comment by amyoshino** [comment]
@ricardoV94 can you restart the tests? I don't think it is failing due to the changes we've made in this PR. The failed ones below seems to be unrelated with the icdf functions, let me know if I am mistaken. 

tests/distributions/test_mixture.py::TestHurdleMixtures::test_hurdle_zero_draws[dist-0.05-non_psi_args8] FAILED [ 97%]
tests/distributions/test_mixture.py::TestHurdleMixtures::test_hurdle_zero_draws[dist-0.3-non_psi_args9] FAILED [ 97%]
tests/distributions/test_mixture.py::TestHurdleMixtures::test_hurdle_zero_draws[dist-0.4-non_psi_args10] FAILED [ 97%]
tests/distributions/test_mixture.py::TestHurdleMixtures::test_hurdle_zero_draws[dist-0.9-non_psi_args11] FAILED [ 98%]

**Comment by ricardoV94** [comment]
They could actually be related. The hurdle mixture uses truncated rvs for the non-zero part. And with these icdfs several truncated variables will switch from rejection sampling to inverse cdf sampling

**Comment by amyoshino** [comment]
Got it, thanks for the pointers. I am still investigating what is the problem. 
I have spent few hours trying to debug but I still can't find what is going on. 
So far, drawing from `Gamma` works well, but inside the `Truncated` it does not. The same do not happen with other functions that I tried (Normal, Beta, HalfCauchy, InverseGamma - picked here some that have icdf functions and some that does not).

It looks like the error points to this numpy function, and now I am in the quest to find how to access `high` and `low` for the Truncated Gamma: https://github.com/numpy/numpy/blob/448e4dabb5e7d593af7d164a335dc7e44e17ec95/numpy/random/_generator.pyx#L1081

Here is what I got until now, and I will keep you posted about progress. But If by looking at it you have any clues, let me know. 

Drawing from Gamma, Beta, etc works as expected (I added some prints for debug purposes):
image

Drawing from Truncated only fails for Gamma:
image

**Comment by amyoshino** [comment]
I have an update but it is still a mystery for me that it raises the `high - low overflow error` when passing only lower but not when passing lower and upper. And that it only happens with `Truncated(Gamma)` but not with `Truncated(Beta)` for example.

image

**Comment by amyoshino** [comment]
I just narrowed the problem down to the following:
As stated by @ricardoV94, the hurdle mixture uses truncated rvs for the non-zero part. The failing tests are breaking when passing Gamma.dist to the Truncated.dist. 
- When I try to simulate the part it breaks the test with `Truncated.dist(Gamma.dist(alpha = 5, beta = 2), lower = 0.01).eval()` it  fails in the middle when trying to sample from a uniform distribution with `UniformRV.rng_fn`. It does so because the inputs for this function (rng_fn) in `tensor.random.op.py` is something like [array(7.64934096e-07), array(nan)] and when trying to generate a number with upper or lower as NaN, it gives the Overflow error we are observing (see [this link line 1081](https://github.com/numpy/numpy/blob/448e4dabb5e7d593af7d164a335dc7e44e17ec95/numpy/random/_generator.pyx#L1081))
- When trying another distribution with icdf implemented, like Beta.distr, and run `Truncated.dist(Beta.dist(alpha = 5, beta = 2), lower = 0.01).eval()` it works. In this case the arguments for the rng_fn function is something like [array(1.835008e-05), array(1.)], the number is then sampled from a uniform distribution and the function finish running. 
- Another experiment that went successful is when I pass lower and upper to `Truncated.dist(Gamma.dist(alpha = 5, beta = 2), lower = 0.01, upper = 1000).eval()` it finishes without any problems. 

I am clueless about why in Gamma the Op does not create the second argument in the case of Truncated(Gamma) but does for other Truncated distributions I have tested. I am stuck in the part where it fails when calling the function`getattr()` inside `pytensor.tensor.random.op.rng_fn()`, and running the debug mode in this goes to so many places that makes me lost. In summary, I did my best in trying to find where (and if) I can solve this problem within the scope of this PR, and can't proceed here without help with this.

**Comment by ricardoV94** [comment]
Is the nan coming from `pm.logcdf(pm.Gamma.dist(alpha=2, beta=5), np.inf)?

**Comment by ricardoV94** [comment]
Sounds like PyTensor gammainc C code does not have a special case for np.inf:

https://github.com/pymc-devs/pytensor/blob/19e1a982e5a2e9ba2bd9d708133abf102c6fb835/pytensor/scalar/c_code/gamma.c#L217-L223

In contrast this scipy has a special case? 
https://github.com/scipy/scipy/blob/4ecf5638c36c3a87b473056f5fffecfeddffd368/scipy/special/cephes/igam.c#L149-L151

**Comment by amyoshino** [comment]
> `pm.logcdf(pm.Gamma.dist(alpha=2, beta=5), np.inf)`

Indeed, this code returns nan. I tried to fix it writing the special case for np.inf in the `gamma.c` function but it didn't work out yet. Maybe there are other places I need to change in order to make it work. 


`DEVICE double GammaP (double n, double x)
{                               /* --- regularized Gamma function P */
  if ((n <= 0) || (x < 0)) return NPY_NAN;  /* check the function arguments */
  if (x <=  0) return 0;        /* treat x = 0 as a special case */
  if (x < n+1) return _series(n, x) *exp(n *log(x) -x -logGamma(n));
  if (isinf(n)) {
    if (isinf(x)) return NPY_NAN;
    return 0;
  } 
  if (isinf(x)) return 1;
  return 1 -_cfrac(n, x) *exp(n *log(x) -x -logGamma(n));
}  /* GammaP() */

/*--------------------------------------------------------------------*/

DEVICE double GammaQ (double n, double x)
{                               /* --- regularized Gamma function Q */
  if ((n <= 0) || (x < 0)) return NPY_NAN;  /* check the function arguments */
  if (x <=  0) return 1;        /* treat x = 0 as a special case */
  if (x < n+1) return 1 -_series(n, x) *exp(n *log(x) -x -logGamma(n));
  if (isinf(n)) {
    if (isinf(x)) return NPY_NAN;
    return 1;
  } 
  if (isinf(x)) return 0;
  return _cfrac(n, x) *exp(n *log(x) -x -logGamma(n));
}  /* GammaQ() */`

Even with this, the `pm.logcdf(pm.Gamma.dist(alpha=2, beta=5), np.inf)` still returns nan.

**Comment by ricardoV94** [comment]
You need to delete pytensor cache or bump the Op c_cache_version number. Try `pytensor-cache clear` or `pytensor cache-clear` I never remember. If clear doesn't work use purge

**Comment by ricardoV94** [comment]
That PR should probably do a similar fix for the complementary gammainc function (gammaincc?)

**Comment by ricardoV94** [comment]
@amyoshino PyTensor new release should be in conda soon. We will need to update the pin on PyTensor version here, as it was a "major" release (or in a separate PR if there are other things that need fix) 

https://github.com/pymc-devs/pytensor/releases/tag/rel-2.19.0

**Comment by amyoshino** [comment]
> @amyoshino PyTensor new release should be in conda soon. We will need to update the pin on PyTensor version here, as it was a "major" release (or in a separate PR if there are other things that need fix)
> 
> https://github.com/pymc-devs/pytensor/releases/tag/rel-2.19.0

That's awesome!! Looking forward to it to be updated! 
Let me know if I need to do anything with respect to this PR as part of the process.

**Comment by ricardoV94** [comment]
You need to bump the PyTensor dependency constraints on the PyMC side

**Comment by amyoshino** [comment]
Ah, I see! Alright, let me try that then! Thanks for the pointer.

**Comment by amyoshino** [comment]
Just pushed the bump dependency modifications in here. Seems like there is one test failing though, which wasn't failing when I ran the test suite locally.

**Comment by ricardoV94** [comment]
@amyoshino I'm taking core of the Pytensor bump in https://github.com/pymc-devs/pymc/pull/7203

**Comment by ricardoV94** [comment]
> Just pushed the bump dependency modifications in here. Seems like there is one test failing though, which wasn't failing when I ran the test suite locally.

May have been an unlucky seed, rerunning it

**Comment by amyoshino** [comment]
> @amyoshino I'm taking core of the Pytensor bump in #7203

In much better hands now 😄 


> > Just pushed the bump dependency modifications in here. Seems like there is one test failing though, which wasn't failing when I ran the test suite locally.
> 
> May have been an unlucky seed, rerunning it

It was indeed a bad luck, the only missing part was the mypy, I guess because of the very outdated repo. I rebased and left only the previous commits here, so it is failing now because PyTensor is still 2.18. I thin the way then is to wait for your PR to be merged and try again, right?

**Comment by ricardoV94** [comment]
> I think the way to go then is to wait for your PR to be merged and try again, right?

Right. I was hoping it was an easier bump, but it also unpins a Python major version :D

**Comment by amyoshino** [comment]
Cool! It looks like the bump is merged! Looking forward to restart the checks when you think is appropriate.

**Comment by amyoshino** [comment]
> > Cool! It looks like the bump is merged! Looking forward to restart the checks when you think is appropriate.
> 
> You can rebase anytime to incorporate the bump

Ah, true! I got so excited that the bump was merged and this potentially can be merged after so long that I forgot that 😅  My apologies haha

**Review by ricardoV94 — APPROVED** [review]
Amazing work @amyoshino !!

**Comment by twiecki** [comment]
Congrats, this was a tricky one!

**Comment by amyoshino** [comment]
That was fun! 😄 
Thank you for the support through this journey!

**Pull request merged**
